Use custom exception type for out of bound in Stack

mihails-strasuns-sociomantic opened the pull request:
Fixes #297

mihails-strasuns-sociomantic committed:
Use custom exception type for out of bound in Stack

Fixes #297

For smoother migration purpose new exception type is still derived
from ArrayBoundsException in D1 (== D2 RangeError). In D2 it is derived
from plain Exception.

nemanja-boric-sociomantic commented:
I'm scared to say anything about breaking the code cause the different exception is now thrown on this one :no_mouth:

nemanja-boric-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -133,7 +133,7 @@ struct Stack (V, int Size = 0)
                           if (depth < stack.length)
```
These all `if (...)` can be converted to enforce for a nicer code.

mihails-strasuns-sociomantic replied:
Full rewrite of module code style is my next step :)

mihails-strasuns-sociomantic commented:
Oh, you mean our D1-only apps doing `catch (RangeException)` with previous code?

nemanja-boric-sociomantic commented:
Yeah.

mihails-strasuns-sociomantic commented:
This is actually both totally legit and damn scary concern :/

nemanja-boric-sociomantic commented:
I wonder if we can keep throwing `ArrayBoundsException` (`StackException: ArrayBoundsException`) in D1 and in D2 do `StackException: Exception`. Given that the users couldn't (yeah, right) catch `RangeError` in the past, we could get away with this approach.

mihails-strasuns-sociomantic commented:
Sounds on the edge of acceptable :)

nemanja-boric-sociomantic commented:
What's happening with the codecov? Is it not attached, is it #298 a cause for it? I doubt it, but IIRC we had some issues with it in #230, that's why it got too complicated and forgotten in the end.

nemanja-boric-sociomantic commented:
It looks like it's just not reported to the PR: https://codecov.io/github/sociomantic-tsunami/ocean/commit/77b6c23bc60e9b3c23dafe5c68a48989e8bb2c8e

mihails-strasuns-sociomantic commented:
Not entirely sure. If I go their website and open ocean PR list in the reports, I can find it there (and log shows it gets submitted without problems). But it doesn't set any status.

mihails-strasuns-sociomantic commented:
Updated with coding style and tests

mihails-strasuns-sociomantic committed:
Fix evaluation order UB in Stack module

rotateLeft/rotateRight functions have been using postfix pointer
increment as part of assignment expression. In D2 effective order of
evaluation is different and specification does not define it.

mihails-strasuns-sociomantic commented:
Updated: fix another undefined behaviour bug found by added tests :)

gavin-norman-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -311,6 +306,48 @@ struct Stack (V, int Size = 0)
+    }
+}
+
+// HACK: some D1 code may be trying to catch ArrayBoundsException specifically
```
Would be worth mentioning what the plan is for removing this hack.

mihails-strasuns-sociomantic replied:
On this topic - any idea how we could possibly handle it? I can only think of deprecating `ArrayBoundsException` alias in `ocean.core` to ensure no one refers to it directly.

nemanja-boric-sociomantic replied:
I think that's fair enough.

gavin-norman-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -311,6 +306,48 @@ struct Stack (V, int Size = 0)
+
+/*******************************************************************************
+
+    Exception used instead of uncatchable `RangeError`.
```
This comment only seems relevant from the perspective of this moment in time (i.e. this commit). It doesn't seem relevant doc for this module, in general.

gavin-norman-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -284,17 +290,6 @@ struct Stack (V, int Size = 0)
-
-        **********************************************************************/
-
-        private V error (size_t line)
```
What? Good riddance.

gavin-norman-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -203,7 +207,8 @@ struct Stack (V, int Size = 0)
                    return p[1] = v;
                    }
 
-                return error (__LINE__);
```
`return`? This code was super weird!

gavin-norman-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -1,309 +1,316 @@
 /*******************************************************************************
```
>Fix Stack coding style and cover it with tests

In a patch release?

mihails-strasuns-sociomantic replied:
This module had exactly 0 tests so I had to do something to actually check that things work. Now that they do, we could possibly move all the part adding tests & style to minor but it would be a bit scary to merge fix without tests coming in a pair.

gavin-norman-sociomantic replied:
The massive coding style changes are more what I was referring to.

mathias-lang-sociomantic commented on src/ocean/util/container/more/Stack.d:
```diff
@@ -351,43 +429,43 @@ static this ( )
-*******************************************************************************/
+ *******************************************************************************/
 
 debug (Stack)
```
Remove ?

mihails-strasuns-sociomantic replied:
Already removed in 3.5.x, don't want conflicts

mathias-lang-sociomantic replied:
But you changed the indent, so you'll get them anyway

mihails-strasuns-sociomantic committed:
Add tests for Stack container

mihails-strasuns-sociomantic commented:
Modified PR to only contains fixes and added tests. Will make another one against v3.x.x after point release with the rest.

mihails-strasuns-sociomantic committed:
Fix void initialization UB bug in Stack module

Initializing stack struct to `void` and trying to access it field right
after is extremely naughty.

mihails-strasuns-sociomantic commented:
Pushed one more UB fix. This module is about to overload my crap detector.

mihails-strasuns-sociomantic commented:
Green
